### PR TITLE
Add GOV.UK Notify as a proper external system

### PR DESCRIPTION
**Pull request by Floppy**
## What does this pull request do?

Adds GOV.UK Notify as a proper external system. Affected diagrams are shown below. I'm not completely happy with the user relationship diagram for Book Video Link, because you have to include Notify in diagrams where it's not really relevant in order to make the relationships that go through it show up. Any ideas on that welcome.

## What is the intent behind these changes?

To properly represent that system in our diagrams

## Changed diagrams

<details>
  <summary>System Overview</summary>

structurizr-56937-system-overview
Notify is actually off the edge of this preview image, but it's there and will show up in structurizr.
</details>

<details>
  <summary>Book Video Link system context</summary>

structurizr-56937-bookVideoLinkSystemContext  
</details>

<details>
  <summary>Book Video Link container</summary>

structurizr-56937-bookVideoLinkContainer
</details>

<details>
  <summary>Book Video Link user relationships</summary>

structurizr-56937-bookVideoLinkUserRelationships
</details>

<details>
  <summary>Interventions system context</summary>

structurizr-56937-interventions-context
</details>

<details>
  <summary>Interventions container</summary>

structurizr-56937-interventions-container
</details>

<details>
  <summary>Mange a Supervision system context</summary>

structurizr-56937-manage-a-supervision-context
</details>

<details>
  <summary>Mange a Supervision container</summary>

structurizr-56937-manage-a-supervision-container
</details>

<details>
  <summary>Mange POM Cases system context</summary>

structurizr-56937-manage-POM-cases-context
</details>

<details>
  <summary>Mange POM Cases container</summary>

structurizr-56937-manage-POM-cases-container
</details>

### CODE REVIEW

**Comment by Floppy** [comment]
One problem here is that if the only link to a `Person` is via notify, it doesn't get pulled in to the context diagram. Will try to fix that.

**Comment by rj-adams** [comment]
Not necessarily just for this PR but could we clarify what "OutsideHMPPS" actually means - is it something (a service) which is:

1) used by HMPPS but not supported or owned by HMPPS (e.g. gov.Notify)

and/or

2) used and owned by HMPPS but not supported by HMPPS (e.g. Curious)

and/or

3) something else ...

**Comment by Floppy** [comment]
I would think of it as 1, but 2 might be another useful category to be able to annotate.

File: src/main/kotlin/model/Interventions.kt [hunk 1]
```diff
@@ -98,8 +98,9 @@ class Interventions private constructor() {
       InterventionTeams.crsProvider.uses(ui, "maintains directory and delivery of dynamic framework interventions and services in")
       ProbationPractitioners.nps.uses(ui, "refers and monitors progress of their service users' interventions and services in")
 
-      service.delivers(InterventionTeams.crsProvider, "emails new referrals", "gov.uk notify")
-      service.delivers(ProbationPractitioners.nps, "emails attendance and safeguarding issues", "gov.uk notify")
+      service.uses(Notify.system, "delivers notifications via")
+      Notify.system.delivers(InterventionTeams.crsProvider, "emails new referrals", "email")
+      Notify.system.delivers(ProbationPractitioners.nps, "emails attendance and safeguarding issues", "email")
```
Comment (sldblog) [inline src/main/kotlin/model/Interventions.kt:103]: I considered this earlier, but going through a service loses the connection between interventions and the actors

So, practically, finding out which service sends CRS providers emails becomes impossible

I think for us, that service-user link has more value than having notify as an external system

Comment (sldblog) [reply src/main/kotlin/model/Interventions.kt:103]: Adding `service.uses(Notify.system, "delivers notifications via")` on top of the original would work though

So perhaps this is an abstraction of "intermediary tool", something like
```
Notify.delivers(from = service, to = InterventionTeams.crsProvider, label = "emails new referrals")
```
and then could register both?

Comment (Floppy) [reply src/main/kotlin/model/Interventions.kt:103]: That sounds sensible - I agree that losing that link would be a real pain and make some very messy diagrams, but in terms of identifying data flows, threat models, etc, notify should be treated as a proper system. Not quite sure how your intermediary tool approach works as I'm pretty new to this repo/structure, but it sounds sensible from what you've said.

Comment (sldblog) [reply src/main/kotlin/model/Interventions.kt:103]: Sorry, bad choice of words from my part (and lots of delay... sorry)

I meant like a DSL, similar to Outside HMPPS:
```
class OutsideHMPPS {
  companion object : addTo {
    override fun addTo(element: Element) {
      when (element) {
        is SoftwareSystem -> element.setLocation(Location.External)
        is Person -> element.setLocation(Location.External)
      }
      {snip}
```

**Comment by swilson-mojhub** [comment]
Worth adding that the Check my Diary Service also uses GOV.UK Notify. 

Also not fused how we illustrate that but having that data is useful

**Comment by sldblog** [comment]
#143 adds the same functionality via a wrapper class
closing as that PR introduces a gov.uk notify system based on this PR

**Pull request closed without merging**